src/intents.js

```diff
@@ -27,36 +25,86 @@ function injectService (url, element, intent) {
 
       if (event.data === 'intent:ready') {
         handshaken = true
-        return event.source.postMessage(intent.data, event.origin)
+        return event.source.postMessage(data, event.origin)
       }
 
       window.removeEventListener('message', messageHandler)
+      iframe.parentNode.removeChild(iframe)
 
       if (event.data === 'intent:error') {
         return reject(new Error('Intent error'))
       }
 
       return handshaken
-        ? iframe.parentNode.removeChild(iframe) && resolve(event.data)
+        ? resolve(event.data)
         : reject(new Error('Unexpected handshake message from intent service'))
     }
 
     window.addEventListener('message', messageHandler)
   })
 }
 
-export function start (client, intent, element, temporaryMockedIntentUrl) {
-  ['action', 'type'].forEach((property) => {
-    if (!intent[property]) {
-      throw new Error(`Misformed intent, "${property}" property must be provided`)
+export function create (cozy, action, type, data = {}, permissions = []) {
+  if (!action) throw new Error(`Misformed intent, "action" property must be provided`)
+  if (!type) throw new Error(`Misformed intent, "type" property must be provided`)
+
+  const createPromise = cozyFetchJSON(cozy, 'POST', '/intents', {
+    data: {
+      type: 'io.cozy.intents',
+      attributes: {
+        action: action,
+        type: type,
+        data: data,
+        permissions: permissions
+      }
     }
   })
 
-  // To replace with a call to client
-  return getUrl(intent, temporaryMockedIntentUrl)
-    .then(url => injectService(url, element, intent))
+  createPromise.start = (element) => {
+    return createPromise.then(intent => {
+      let service = intent.attributes.services && intent.attributes.services[0]
+
+      if (!service) {
+        return Promise.reject(new Error('Unable to find a service'))
+      }
+
+      return injectService(service.href, element, data)
+    })
+  }
+
+  return createPromise
+}
+
+function listenClientData (intent, window) {
+  return new Promise((resolve, reject) => {
+    const messageEventListener = (event) => {
+      if (event.origin !== intent.attributes.client) return
+
+      window.removeEventListener('message', messageEventListener)
+      resolve(event.data)
+    }
+
+    window.addEventListener('message', messageEventListener)
+    window.parent.postMessage('intent:ready', intent.attributes.client)
+  })
 }
 
-export function resolve () {
-  // TODO
+// returns a service to communicate with intent client
+export function createService (cozy, id, window) {
+  return cozyFetchJSON(cozy, 'GET', `/intents/${id}`)
+    .then(intent => {
+      return listenClientData(intent, window)
+        .then(data => {
```

Comment:
This whole thing is really nice, but is there any particular reason why you don't chain the promises instead of nesting them? ie.

```
export function createService (cozy, id, window) {
  return cozyFetchJSON(cozy, 'GET', `/intents/${id}`)
    .then(intent => {
      return listenClientData(intent, window)
    })
    .then(data => {
          let terminated = false
          return etc...
     })
}
```